pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -51,21 +55,83 @@ type ReaderPool struct {
 	lazyReaders   map[*LazyBinaryReader]struct{}
 }
 
-// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers if enabled.
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+type HeadersLazyLoaded struct {
+	Path   string
+	UserID string
+}
+
+type HeadersLazyLoadedTrackerState struct {
+	LazyLoadedBlocks map[string]int64 `json:"lazy_loaded_blocks"`
+	UserID           string           `json:"user_id"`
+}
+
+func (p *ReaderPool) persist(state HeadersLazyLoadedTrackerState, finalPath string) error {
+	// Create temporary path for fsync
+	tmpPath, err := os.CreateTemp("", "lazy-loaded")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpPath.Name())
+
+	data, err := json.MarshalIndent(state, "", "")
+	if err != nil {
+		return err
+	}
+
+	merr := multierror.New()
+	_, err = tmpPath.Write(data)
+	merr.Add(err)
+	merr.Add(tmpPath.Sync())
+	merr.Add(tmpPath.Close())
+
+	if err := merr.Err(); err != nil {
+		return err
+	}
+
+	// fsync the directory too
+	dir, err := os.OpenFile(path.Dir(tmpPath.Name()), os.O_RDONLY, 0777)
+	if err != nil {
+		return err
+	}
+
+	merr.Add(dir.Sync())
+	merr.Add(dir.Close())
+
+	// move the written file to the actual path
+	return os.Rename(tmpPath.Name(), finalPath)
+}
+
+// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers and blockIds writers if enabled.
+func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics, headersLazyLoaded HeadersLazyLoaded) *ReaderPool {
 	p := newReaderPool(logger, lazyReaderEnabled, lazyReaderIdleTimeout, metrics)
 
 	// Start a goroutine to close idle readers (only if required).
 	if p.lazyReaderEnabled && p.lazyReaderIdleTimeout > 0 {
 		checkFreq := p.lazyReaderIdleTimeout / 10
 
 		go func() {
+			tickerLazyLoad := time.NewTicker(time.Minute)
+			defer tickerLazyLoad.Stop()
+
+			tickerIdleReader := time.NewTicker(checkFreq)
+			defer tickerIdleReader.Stop()
+
 			for {
 				select {
 				case <-p.close:
 					return
-				case <-time.After(checkFreq):
+				case <-tickerIdleReader.C:
 					p.closeIdleReaders()
+				case <-tickerLazyLoad.C:
+					state := HeadersLazyLoadedTrackerState{
+						LazyLoadedBlocks: p.LoadedBlocks(),
+						UserID:           headersLazyLoaded.UserID,
+					}
+
+					// Then we persist the state to files so that we are not holding lock for too long.
```

Comment:
This comment seems a bit out outdated here, there is no lock holding at least in this function.